src/libraries/System.Text.Json/source_generation/src/JsonSourceGeneratorHelper.Generate.cs

```diff
@@ -0,0 +1,672 @@
+
+            return @$"{GetUsingStatementsString(typeMetadata)}
+
+namespace {_generationNamespace}
+{{
+    {JsonContextDeclarationSource}
+    {{
+        private JsonTypeInfo<{typeCompilableName}> _{typeFriendlyName};
+        public JsonTypeInfo<{typeCompilableName}> {typeFriendlyName}
+        {{
+            get
+            {{
+                if (_{typeFriendlyName} == null)
+                {{
+                    {WrapWithCheckForCustomConverterIfRequired(metadataInitSource, typeCompilableName, typeFriendlyName, GetNumberHandlingAsStr(typeMetadata.NumberHandling))}
+                }}
+
+                return _{typeFriendlyName};
+            }}
+        }}{additionalSource}
+    }}
+}}
+";
+        }
+
+        private string WrapWithCheckForCustomConverterIfRequired(string source, string typeCompilableName, string typeFriendlyName, string numberHandlingNamedArg)
+        {
+            if (!_honorRuntimeProvidedCustomConverters)
+            {
+                return source;
+            }
+
+            return @$"JsonConverter customConverter;
+                    if ({OptionsInstanceVariableName}.Converters.Count > 0 && (customConverter = {RuntimeCustomConverterFetchingMethodName}(typeof({typeCompilableName}))) != null)
+                    {{
+                        _{typeFriendlyName} = {JsonMetadataServicesClassName}.{GetCreateValueInfoMethodRef(typeCompilableName)}({OptionsInstanceVariableName}, customConverter);
+                    }}
+                    else
+                    {{
+                        {source.Replace(Environment.NewLine, $"{Environment.NewLine}    ")}
+                    }}";
+        }
+
+        // Base source generation context partial class.
+        private string BaseJsonContextImplementation()
+        {
+            StringBuilder sb = new();
+            sb.Append(@$"using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace {_generationNamespace}
+{{
+    {JsonContextDeclarationSource}
+    {{
+        private static JsonContext s_default;
+        public static JsonContext Default => s_default ??= new JsonContext(new JsonSerializerOptions());
+
+        public JsonContext() : base(null)
+        {{
+        }}
+
+        public JsonContext(JsonSerializerOptions options) : base(options)
+        {{
+        }}
+
+        {GetFetchLogicForRuntimeSpecifiedCustomConverter()}
+    }}
+}}
+");
+
+            return sb.ToString();
+        }
+
+        private string GetFetchLogicForRuntimeSpecifiedCustomConverter()
+        {
+            if (!_honorRuntimeProvidedCustomConverters)
+            {
+                return "";
+            }
+
+            return @$"private JsonConverter {RuntimeCustomConverterFetchingMethodName}(System.Type type)
+        {{
+            System.Collections.Generic.IList<JsonConverter> converters = {OptionsInstanceVariableName}.Converters;
+
+            for (int i = 0; i < converters.Count; i++)
+            {{
```

Comment:
Add a "todo: use a dictionary if count > x" where `x` is probably in the 10-20 range.